### PR TITLE
(PUP-10053) Update to more recent webmock and vcr

### PR DESCRIPTION
**Pull request by joshcooper**
Update to more recent versions for webmock and vcr, both of which dropped
support for older ruby versions.

Rewrite tests to use `stub_request` with request headers, basic_auth, and use
built-in functionality for stubbing responses. Also drop places where we were
using `allow|expect_any_instance_of(Net::HTTP)`.

Due to the way webmock is configured, the stubs act like expectations,
because webmock will raise an exception if a request is made that
doesn't match the stub.

Drop unneeded VCR cassettes for `Connection#request_*` methods.

Only enable VCR for tests which enable it using `:vcr => true`

### CODE REVIEW

**Comment by joshcooper** [comment]
Supersedes PR #7767 7744

**Comment by joshcooper** [comment]
@jtappa I retargeted this PR at 6.4.x and added your comment about debugging VCR.

File: spec/unit/network/http/connection_spec.rb [hunk 1]
```diff
@@ -138,164 +119,131 @@ def expects_limit_exceeded(conn)
...
-      pool = expects_redirection(conn)
-      expect(pool).to receive(:with_connection).with(other_site, anything).exactly(3).times
+      stub_request(:get, "http://me.example.com:8140/").to_return(redirect_to("http://me.example.com:8140/foo"))
+      stub_request(:get, "http://me.example.com:8140/foo").to_return(redirect_to("http://me.example.com:8140/bar"))
+      stub_request(:get, "http://me.example.com:8140/bar").to_return(redirect_to("http://me.example.com:8140/baz"))
+      stub_request(:get, "http://me.example.com:8140/baz").to_return(redirect_to("http://me.example.com:8140/qux"))
 
+      conn = create_connection(:redirect_limit => 3)
       expects_limit_exceeded(conn)
     end
```
Comment (kris-bosland) [inline spec/unit/network/http/connection_spec.rb]: I would like to see a test where there is a redirection limit, but the limit is not exceeded, and the connection is OK.

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -155,6 +161,24 @@ module PuppetSpec
     Puppet::Test::TestHelper.before_each_test()
   end
 
+  config.around :each do |example|
+    # Ignore requests from Facter GCE fact in Travis
+    stub_request(:get, "http://metadata/computeMetadata/v1beta1/?alt=json&recursive=true") # facter 2
+    stub_request(:get, "http://metadata.google.internal/computeMetadata/v1/?recursive=true&alt=json") # facter 3
```
[the remark this answers is not in the document]
Comment (joshcooper) [reply spec/spec_helper.rb]: If facter changes the GCE endpoint then new PRs to puppet will begin to fail, so we'll know to update puppet.

File: spec/unit/network/http/connection_spec.rb (hunk 1 above)
Comment (joshcooper) [reply spec/unit/network/http/connection_spec.rb]: That's tested in https://github.com/puppetlabs/puppet/pull/7767/files/1c63f9d3231128ceac96fbee00a474942b67ca67#diff-f12903d4e996470e4f3649e6a5d6eb5eR109, as the default redirect limit is 10

**Pull request merged**